jsonparser/tidwall.go

```diff
@@ -0,0 +1,253 @@
+package jsonparser
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/samber/lo"
+
+	"github.com/rudderlabs/rudder-go-kit/jsonrs"
+
+	"github.com/tidwall/gjson"
+	"github.com/tidwall/sjson"
+)
+
+// tidwallJSONParser is the implementation of JSONParser using gjson/sjson libraries
+type tidwallJSONParser struct{}
+
+// GetValue retrieves the value for a given key from JSON bytes using gjson
+func (p *tidwallJSONParser) GetValue(data []byte, keys ...string) (interface{}, error) {
+	if len(data) == 0 {
+		return nil, ErrEmptyJSON
+	}
+
+	// Handle empty keys or no keys - return the entire JSON object
+	if len(keys) == 0 || (len(keys) == 1 && keys[0] == "") {
+		var result interface{}
+		if err := jsonrs.Unmarshal(data, &result); err != nil {
+			return nil, fmt.Errorf("failed to unmarshal JSON: %w", err)
+		}
+		return result, nil
+	}
+
+	// Use gjson to get the value
+	result := gjson.GetBytes(data, getPath(keys...))
+	if !result.Exists() {
+		return nil, fmt.Errorf("key not found: %s", keys)
+	}
+
+	// Convert the gjson.Result to the appropriate Go type
+	switch result.Type {
+	case gjson.String:
+		return result.String(), nil
+	case gjson.Number:
+		return result.Float(), nil
+	case gjson.True:
+		return true, nil
+	case gjson.False:
+		return false, nil
+	case gjson.Null:
+		// nolint: nilnil
+		return nil, nil
+	case gjson.JSON:
+		if result.IsArray() {
+			var arr []interface{}
+			if err := jsonrs.Unmarshal([]byte(result.Raw), &arr); err != nil {
```

Comment:
Why not return `[]byte`? It feels like we're doing too much in this package already. Also, this won't allow us to run `GetValue` again on the returned result.

Consider this JSON:
```json
{
	"name": "John",
	"surname": "Doe",
	"age": 30,
	"city": {
		"name": "New York",
		"zipcode": "10001",
		"country": "USA"
	}
}
```

I could get `city` and then on the returned result (if it were a slice of bytes), I could call `GetValue` again multiple times to get `name`, `zipcode` and `country`.

Reply:
If possible I'd try to remove all traces of `json.Unmarshal` from this package.